File: ruby/hamming/hamming.rb
module Hamming
  VERSION = 1

  class << self
    def compute(a, b)
      fail ArgumentError unless a.length == b.length

      char_pairs(a, b)
        .map { |x, y| x == y }
        .count(false)
    end

    private

    def char_pairs(a, b)
      a.each_char.zip(b.each_char)
    end
  end
end

File: elixir/hamming/hamming.exs
defmodule DNA do
  @doc """
  Returns number of differences between two strands of DNA, known as the Hamming Distance.

  ## Examples

  iex> DNA.hamming_distance('AAGTCATA', 'TAGCGATC')
  4
  """
  @spec hamming_distance([char], [char]) :: non_neg_integer
  def hamming_distance(s1, s2) when length(s1) != length(s2), do: nil
  def hamming_distance(strand1, strand2) do
    Enum.zip(strand1, strand2)
    |> Enum.count(fn { s1, s2 } -> s1 != s2 end)
  end
end
